social-tw/social-tw-website
# [Relay] [UST-171] [UST-172] Feature/ust 171 claim positive reputation api

## Summary

Add Claim Negative/Positive Reputation (Relayer) & API

## Linked Issue

close #384 #385 

## Details

Add claim Negative/Positive Reputation (Relayer) & API

## Impacted Areas

`packages/relay/src/routes/reportsRoute.ts`
`packages/relay/src/services/ReportService.ts`
`packages/relay/src/services/utils/IpfsHelper.ts`
`packages/relay/src/types/InternalError.ts`
`packages/relay/src/types/Reputation.ts`
`packages/relay/test/report.test.ts`
`packages/relay/test/reputation_claim.test.ts`

## Tests

- Added unit tests for positive and negative reputation claims
- Included tests for different user types (reporter, poster, voter)
- Added tests for error cases 
- Updated existing tests to match new response format

## Possible Impacts

Describe any existing features that might be affected by this change, or any potential backward compatibility issues, if any.

## Visual Materials

N/A (This change doesn't affect the UI)

## Verification Steps

1. Start the relay server
2. Use the API to create a report
3. Vote on the report
4. Attempt to claim positive reputation for the reporter
5. Attempt to claim negative reputation for the poster
6. Attempt to claim positive reputation for the voter
7. Verify that appropriate errors are thrown for invalid claims or repeated claims

## Todo

Break down the progress of the PR for everyone to see what else you intend to include in this PR.

- [x] Implement claim reputation functionality in ReportService
- [x] Update API endpoints for claiming reputation
- [x] Improve error handling for reputation claims
- [x] Update reputation status update logic
- [x] Enhance and add new test cases
- [ ] Update API documentation to reflect changes in response format

## Checklist

- [ ] I have updated the documentation (API docs need to be updated)
